README.md
# qsmooth

### Global normalization methods and their assumptions

Global normalization methods such as quantile normalization have become a standard part of the
analysis pipeline for high-throughput data to remove unwanted technical variation. These 
methods and others that rely solely on observed data without external information (e.g. spike-ins)
are based on the assumption that only a minority of genes are expected to be differentially
expressed (or that an equivalent number of genes increase and decrease across biological conditions. 
This assumption can be interpreted in different ways leading to different global normalization 
procedures. For example, in one normalization procedure, the method assumes
the mean expression level across genes should be the same across samples. 
In contrast, quantile normalization assumes the only difference between the statistical distribution
of each sample is technical variation. Normalization is achieved by forcing the observed distributions
to be the same and the average distribution, obtained by taking the average of each quantile 
across samples, is used as the reference. 

### How to evaluate if global normalization methods are appropriate? 

While these assumptions may be reasonable in certain experiments, they may not always be 
appropriate. Recently, an R/Bioconductor package ([`quantro`](http://www.bioconductor.org/packages/release/bioc/html/quantro.html))
has been developed to test for global differences between groups of distributions to evaluate whether 
global normalization methods such as quantile normalization should be applied. If global differences
are found between groups of distributions, these changes may be of technical or biological of interest.
If these changes are of technical interest (e.g. batch effects), then global normalization methods should be applied. 
If these changes are related to a biological factor (e.g. normal/tumor or two tissues), then 
global normalization methods should not be applied because the methods will remove the interesting biological variation 
(i.e. differentially expressed genes) and artificially induce differences between genes that were not 
differentially expressed. In the cases with global differences between groups of distributions 
between biological conditions, quantile normalization is not an appropriate normalization method. In 
these cases, we can consider **a more relaxed assumption** about the data, namely that the statistical distribution
of each sample should be the same within biological conditions or groups (compared to the more 
stringent assumption of quantile normalization, which states the statistical distribution is the same across all samples).

### qsmooth: a generalization of quantile normalization

Here we introduce a generalization of quantile normalization, referred to as `smooth quantile normalization`
(**qsmooth**), which is a weighted average of the two types of assumptions about the data. 
The **qsmooth** R-package contains the `qsmooth()` function, which computes a weight at every quantile
that compares the variability between groups relative to within groups. In one extreme, quantile normalization 
is applied and in the other extreme quantile normalization within each biological condition is applied.
The weight shrinks the group-level quantile normalized data towards the overall reference quantiles 
if variability between groups is sufficiently smaller than the variability within groups. The algorithm is described in the
Figure below (see the [`vignettes/qsmooth-vignette.pdf`](https://github.com/stephaniehicks/qsmooth/blob/master/vignettes/qsmooth-vignette.pdf) for more details). 

![qsmooth algorithm](https://github.com/stephaniehicks/qsmooth/raw/master/vignettes/qsmooth_algo.jpg)

### Installing qsmooth

The R-package **qsmooth** can be installed from Github using the R 
package [devtools](https://github.com/hadley/devtools): 


Use  to install the latest version of **qsmooth** from Github:
```s
library(devtools)
install_github("stephaniehicks/qsmooth")
```

It can also be installed using Bioconductor: 

```s
# install BiocManager from CRAN (if not already installed)
if (!requireNamespace("BiocManager", quietly=TRUE))
    install.packages("BiocManager")

# install qsmooth package
BiocManager::install("qsmooth")
```

After installation, the package can be loaded into R.
```s
library(qsmooth)
```


# Using qsmooth

The main function in the **qsmooth** package is `qsmooth()`.  The `qsmooth()` function needs two objects: 
(1) a data frame or matrix with observations (e.g. probes or genes) on the rows and samples as the columns 
(e.g. let's call it `eset`) and (2) a group level factor called `group_factor` (let's call it `outcome`). 
This order of this factor variable must match the order of the columns in the `eset` object because it contains
 information about which group each sample is from.

To run the `qsmooth()` function, 
```
qs <- qsmooth(object = eset, group_factor = outcome)
```
Individual slots can be extracted using accessor methods:
```
qsmoothData(qs) # extract smoothed quantile normalized data
qsmoothWeights(qs) # extract smoothed quantile normalized weights
```

The weights can be directly plotted using the `qsmoothPlotWeights()` function. 
```
qsmoothPlotWeights(qs) # plot weights 
```

See `vignettes/qsmooth-vignette.pdf` for more details. 

# Bug reports
Report bugs as issues on the [GitHub repository](https://github.com/stephaniehicks/qsmooth)


# Contributors

* [Stephanie Hicks](https://github.com/stephaniehicks)
* [Kwame Okrah](https://github.com/kokrah)


inst/NEWS
Changes in version 0.99.0 (2019-02-07)
+ Submitted to Bioconductor

Changes in version 0.99.2 (2019-03-29)
+ Removed hard-coded date in DESCRIPTION file
+ Included Bioconductor installation instructions in README.md file
+ Changed the name of the vignette so user can use `vigentte("qsmooth")`
+ man. Changed the description of the return to reflect the actual output

Changes in version 0.99.3 (2019-03-29)
+ Created a show method to show only first and last three values in the qsmoothWeights function and the head of the qsmoothData normalized matrix
+ Updated all the examples and vignette to use the ExperimentHub bodymapRat data package 
+ Updated documentation in the plots.R file
+ Included argument checking in the qsmooth.R file. Now checks if object is a matrix, data.frame or SummarizedExperiment. 
+ Included functionality for a user providing a SummarizedExperiment and extracting the counts from the assay slot (if it exists).

Changes in version 0.99.4 (2019-03-29)
+ Running `R CMD BiocCheck` gave me a warning unless I used R version 3.7, but then this failed to pass in qsmooth version 0.99.3. So back to R version 3.6

Changes in version 0.99.5 (2019-03-30)
+ Updating usage examples so R CMD check runs under 5 mins

Changes in version 0.99.6 (2019-03-30)
+ Reducing vignette so R CMD check runs under 5 mins

Changes in version 0.99.7 (2019-03-31)
+ Updating usage examples in qsmooth.R so R CMD check runs under 5 mins

Changes in version 0.99.8 (2019-04-03)
+ Fixed header of vignette .Rmd

Changes in version 0.99.9 (2019-04-04)
+ Adding ORCID for Stephanie Hicks, Hector Corrada Bravo and Rafael Irizarry

Changes in version 0.99.10 (2019-04-04)
+ Fixed typo in ORCID

Changes in version 0.99.11 (2019-04-09)
+ Fixed bug YAML in vignette causing an error on bioc nightly builds

Changes in version 1.3.1 (2020-03-01)
+ Fixed bug on conditionals longer than 1

Changes in version 1.9.1 (2021-06-24)
+ Added qsmoothGC function (Contributed from Koen Van den Berge)

Changes in version 1.9.2. (2021-09-01)
+ Added Hmisc to Imports

Changes in version 1.15.1 (2022-11-04)
+ Removed CC BY 4.0 license and replaced with GPL-3


tests/testthat.R
library(testthat)
context("checking the qsmooth function")

library(qsmooth)

test_that("checking dimensions of qsmooth object", {
  n_genes <- 1000
  n_samples <- 10
  set.seed(1234)
  x <- matrix(rnorm(n_genes*n_samples), n_genes, n_samples)
  group_factor <- rep(c(1,2),each=5)
  qs_norm <- qsmooth(object = x, 
                    group_factor = group_factor)
  
  expect_equal(length(qsmoothWeights(qs_norm)), n_genes)
  expect_equal(nrow(qsmoothData(qs_norm)), n_genes)
  expect_equal(ncol(qsmoothData(qs_norm)), n_samples)
})

test_that("checking qsmooth normalized data", {
  n_genes <- 1000
  n_samples <- 10
  set.seed(1234)
  x <- matrix(rnorm(n_genes*n_samples), n_genes, n_samples)
  group_factor <- rep(c(1,2),each=5)
  qs_norm <- qsmooth(object = x, group_factor = group_factor)
  expect_equal(qsmoothData(qs_norm)[1,1], -1.25856, tolerance = .0001)
  expect_equal(qsmoothWeights(qs_norm)[1], 0.97495, tolerance = .0001)
})


test_that("checking qsmoothGC normalized data", {
  n_genes <- 1000
  n_samples <- 10
  set.seed(1234)
  x <- matrix(rpois(n_genes*n_samples, lambda=12), n_genes, n_samples)
  group_factor <- rep(c(1,2),each=5)
  gc <- runif(n=n_genes, min=0.2, max=0.9)
  ## one group should be same as qsmooth
  qs_norm <- qsmooth(object = x, group_factor = group_factor)
  qs_normGC <- qsmoothGC(object = x, 
                         gc = gc, 
                         group_factor = group_factor, 
                         nGroups = 1,
                         round = FALSE)
  expect_equal(qsmoothData(qs_norm), qsmoothData(qs_normGC), tolerance = .001)
})


R/plots.R
#' @title Plot weights from \code{qsmooth} function.
#' 
#' @description This function plots a scatterplot
#' showing the \code{qsmoothWeights} along the y-axis
#' and the quantiles on the x-axis.
#' 
#' @param object a qsmooth object from \code{qsmooth} 
#' @param xLab label for x-axis. Default is "quantiles"
#' @param yLab label for y-axis. Default is "weights"
#' @param mainLab title of plot. Default is "qsmooth weights"
#' 
#' @return A scatterplot will be created showing the 
#' \code{qsmoothWeights} along the y-axis and the 
#' quantiles on the x-axis.
#' 
#' @importFrom graphics par
#' @importFrom graphics plot
#' @importFrom graphics abline
#' 
#' @export
#' @examples
#' dat <- cbind(matrix(rnorm(1000), nrow=100, ncol=10), 
#'              matrix(rnorm(1000, .1, .7), nrow=100, ncol=10))
#' dat_qs <- qsmooth(object = dat, 
#'                   group_factor = rep(c(0,1), each=10))
#' qsmoothPlotWeights(dat_qs)
#' 
qsmoothPlotWeights <- function(object,
            xLab = "quantiles", yLab = "weights", 
            mainLab = "qsmooth weights"){
    oldpar = par(mfrow=c(1,1), mar=c(4, 4, 1.5, 0.5))
    w = object@qsmoothWeights
    lq = nrow(object@qsmoothData)
    u = (seq_len(lq) - 0.5) / lq
    
    if (length(u) > 1e4) { # do not plot > 10000 points
        sel = sample(seq_len(lq), 1e4)
        plot(u[sel], w[sel], pch=".", main = mainLab,
             xlab = xLab, ylab = yLab, ylim=c(0, 1))
    } else{
        plot(u, w, pch=".", main = mainLab,
             xlab = xLab, ylab = yLab, ylim=c(0, 1))
    }
    abline(h=0.5, v=0.5, col="red", lty=2)
    par(oldpar)
}

R/qstats.R
#' @title qstats 
#'
#' @description This function is a helper function that 
#' computes quantile statistics for the function 
#' \code{qsmooth}. 
#'
#' @param object an object which is a data frame or 
#' matrix with observations (e.g. probes or genes) on 
#' the rows and samples as the columns.  
#' @param group_factor a group level continuous or categorial 
#' covariate associated with each sample or column in the 
#' \code{object}. The order of the \code{group_factor} must 
#' match the order of the columns in \code{object}. 
#' @param window window size for running median which is 
#' a fraction of the number of rows in \code{object}. 
#' Default is 0.05. 
#' 
#' @return A list of quantile statistics including 
#' \item{Q}{sample quantiles}
#' \item{Qref}{reference quantile}
#' \item{Qhat}{linear model fit at each quantile}
#' \item{SST}{total sum of squares}
#' \item{SSB}{between sum of squares}
#' \item{SSE}{within sum of squares}
#' \item{roughWeights}{SSE / SST}
#' \item{smoothWeights}{smoothed weights computed using a 
#' running median with a given \code{window} size.}
#' 
#' @aliases qstats
#' 
#' @docType methods
#' @importFrom stats runmed
#' @importFrom stats model.matrix
#'
#' @examples
#' dat <- cbind(matrix(rnorm(1000), nrow=100, ncol=10), 
#'              matrix(rnorm(1000, .1, .7), nrow=100, ncol=10))
#' qs <- qstats(object = dat, 
#'              group_factor = rep(c(0,1), each=10), 
#'              window = 0.05)
#' 
#' @rdname qstats
#' @export
qstats <- function(object, group_factor, window = 0.05)
  {
  # Compute sample quantiles
  Q = apply(object, 2, sort) 
  
  # Compute quantile reference
  Qref = rowMeans(Q)  
  
  # Compute SST
  SST = rowSums((Q - Qref)^2)
  
  # Compute SSB
  if(is.factor(group_factor)){
    X = model.matrix(~ 0 + group_factor)
  } else {
    X = model.matrix(~ group_factor)
  }
  
  QBETAS = t(solve(t(X) %*% X) %*% t(X) %*% t(Q))
  Qhat = QBETAS %*% t(X)
  SSB = rowSums((Qhat - Qref)^2)
  
  # Compute weights
  roughWeights = 1 - SSB / SST
  roughWeights[SST < 1e-6] = 1
  
  # Compute smooth weights
  k = floor(window * nrow(Q))
  if (k %% 2 == 0) k = k + 1
  smoothWeights = runmed(roughWeights, k = k, endrule="constant")
  
  list(Q=Q, Qref=Qref, Qhat=Qhat, SST=SST, SSB=SSB, SSE=SST-SSB, 
       roughWeights=roughWeights, smoothWeights=smoothWeights)
}



R/methods.R
#' @title Accessors for the 'qsmoothWeights' slot of a qsmooth object.
#' 
#' @description Accessors for the 'qsmoothWeights' slot of a qsmooth object.
#' 
#' @usage
#' \S4method{qsmoothWeights}{qsmooth}(object)
#'
#' @docType methods
#' @name qsmoothWeights
#' @rdname qsmoothWeights
#' @aliases qsmoothWeights qsmoothWeights,qsmooth-method
#' @param object an object of class \code{qsmooth}.
#' 
#' @return The weights calculated for each feature 
#' after applying smoothed quantile normalization. 
#' 
#' @export
#' 
#' @examples
#' dat <- cbind(matrix(rnorm(1000), nrow=100, ncol=10), 
#'              matrix(rnorm(1000, .1, .7), nrow=100, ncol=10))
#' dat_qs <- qsmooth(object = dat, 
#'                   group_factor = rep(c(0,1), each=10))
#' qsmoothWeights(dat_qs)
#' 
setMethod(
  f = "qsmoothWeights", 
  signature = "qsmooth",
  definition = function(object) {
            return(object@qsmoothWeights)
          }
)

#' @title Accessors for the 'qsmoothData' slot of a qsmooth object.
#' 
#' @description Accessors for the 'qsmoothData' slot of a qsmooth object.
#' 
#' @usage
#' \S4method{qsmoothData}{qsmooth}(object)
#'
#' @docType methods
#' @name qsmoothData
#' @rdname qsmoothData
#' @aliases qsmoothData qsmoothData,qsmooth-method
#' @param object an object of class \code{qsmooth}.
#' 
#' @return The normalized data after applying 
#' smoothed quantile normalization. 
#' 
#' @export
#' 
#' @examples
#' dat <- cbind(matrix(rnorm(1000), nrow=100, ncol=10), 
#'              matrix(rnorm(1000, .1, .7), nrow=100, ncol=10))
#' dat_qs <- qsmooth(object = dat, 
#'                   group_factor = rep(c(0,1), each=10))
#' qsmoothData(dat_qs)
#' 
setMethod(
  f = "qsmoothData", 
  signature = "qsmooth",
  definition = function(object) {
            return(object@qsmoothData)
          }
)


R/qsmooth.R
#' @title qsmooth
#' 
#' @description This function applies a generalization of 
#' quantile normalization called smoothed quantile 
#' normalization. This function defines the qsmooth class 
#' and constructor. 
#'
#' @param object an object which is a \code{matrix} or 
#' \code{data.frame} with observations (e.g. probes or genes) on 
#' the rows and samples as the columns. Alternatively, 
#' a user can provide a \code{SummarizedExperiment} object
#' and the \code{assay(object, "counts")} will be used as input 
#' for the qsmooth normalization. 
#' @param group_factor a group level continuous or categorial 
#' covariate associated with each sample or column in the 
#' \code{object}. The order of the \code{group_factor} must 
#' match the order of the columns in \code{object}.
#' @param batch (Optional) batch covariate (multiple 
#' batches are not allowed). 
#' If batch covariate is provided, \code{Combat()} from 
#' \code{sva} is used prior to qsmooth normalization to 
#' remove batch effects. See \code{Combat()} for more details. 
#' @param norm_factors optional normalization scaling factors.
#' @param window window size for running median which is 
#' a fraction of the number of rows in \code{object}. 
#' Default is 0.05. 
#' @export 
#' 
#' @return A object of the class \code{qsmooth} that 
#' contains a numeric vector of the qsmooth weights in 
#' the \code{qsmoothWeights} slot and a matrix of normalized  
#' values after applying smoothed quantile normalization in 
#' the \code{qsmoothData} slot. 
#' 
#' @details 
#' Quantile normalization is one of the most widely used 
#' normalization tools for data analysis in genomics. Although it 
#' was originally developed for gene expression microarrays it is
#' now used across many different high-throughput applications 
#' including RNAseq and ChIPseq. The methodology 
#' relies on the assumption that observed changes in the empirical 
#' distribution of samples are due to unwanted variability. 
#' Because the data is transformed to remove these differences 
#' it has the potential to remove interesting biologically driven
#' global variation. Therefore, applying quantile normalization, 
#' or other global normalization methods that rely on similar 
#' assumptions, may not be an appropriate depending on the type 
#' and source of variation. 
#' 
#' This function computes a weight at every 
#' quantile that compares the variability between groups 
#' relative to within groups. In one extreme quantile 
#' normalization is applied and in the other extreme quantile 
#' normalization within each biological condition is applied. 
#' The weight shrinks the group-level quantile normalized data 
#' towards the overall reference quantiles if variability 
#' between groups is sufficiently smaller than the variability 
#' within groups. 
# 
#' See the vignette for more details. 
#' 
#' @aliases qsmooth
#' 
#' @docType methods
#' @name qsmooth
#' @importFrom sva ComBat
#' @importFrom SummarizedExperiment assays
#' @importFrom SummarizedExperiment assay
#' @importFrom stats ave
#' 
#' @examples
#' dat <- cbind(matrix(rnorm(1000), nrow=100, ncol=10), 
#'              matrix(rnorm(1000, .1, .7), nrow=100, ncol=10))
#' dat_qs <- qsmooth(object = dat, 
#'                   group_factor = rep(c(0,1), each=10))
#' 
#' @rdname qsmooth
#' @export
qsmooth <- function(object, group_factor,
                    batch = NULL, norm_factors = NULL, 
                    window = 0.05)
{

  if(!any(is(object, "matrix") | is(object, "data.frame") |
     is(object, "SummarizedExperiment"))){
    stop("The class of the object must be a matrix, 
         data.frame or SummarizedExperiment")
  }
  
  if(is.data.frame(object)){ object <- as.matrix(object) }
  
  if(is(object, "SummarizedExperiment")){ 
    if("counts" %in% names(assays(object))){ 
      object <- assay(object, "counts")
    } else {
      stop("There is no assay slot named 'counts' inside 
           the object. Please check the names of the 
           assay slots using names(assays(object)).")
    }
  }
  
  if(is.null(group_factor)){  
    stop("Must provide group_factor to specify the group 
        level information associated with each sample or 
             or column in object.")
  }
  
  if(ncol(object) != length(group_factor)){
    stop("Number of columns in object does not match 
             length of group_factor.")
  }
  
  if(is.factor(group_factor) & length(levels(group_factor)) < 2){
    stop(paste0("group_factor is a factor and number of levels in 
                  group_factor is less than 2 (levels(group_factor): ",
                levels(group_factor), "). Must provide a factor with 
                  2 or more levels to use qsmooth."))
  }
  
  if(any(is.na(object))){ stop("Object must not contains NAs.") }    
  
  # Scale normalization step
  if(!is.null(norm_factors)) { 
    object <- sweep(object, 2, norm_factors, FUN = "/") 
  }
  
  # If batch is provided, run sva::ComBat to remove batch effects
  if(!is.null(batch)){
    object <- ComBat(object, factor(batch))
  }
  
  # Compute quantile statistics
  qs <- qstats(object=object, group_factor=group_factor,
               window=window)
  Qref <- qs$Qref 
  Qhat <- qs$Qhat  
  w <- qs$smoothWeights
  
  # Weighted quantiles
  objectNorm = w * Qref + (1 - w) * Qhat
  
  # Re-order objectNorm by rank of object (columnwise)
  for (i in seq_len(ncol(objectNorm))) {
    # Grab ref. i
    ref = objectNorm[,i]
    
    # Grab object column i
    x = object[,i]
    
    # Grab ranks of x (using min rank for ties)
    rmin = rank(x, ties.method="min")
    
    # If x has rank ties then average the values of ref at 
    # those ranks
    dups = duplicated(rmin)
    
    if (any(dups)) {
      # Grab ranks of x (using random ranks for ties) 
      # (needed to uniquely identify the indices of tied ranks)
      rrand = rank(x, ties.method="random")
      
      # Grab tied ranks
      tied.ranks = unique(rmin[dups])
      
      for (k in tied.ranks) {
        # Select the indices of tied ranks 
        sel = rrand[rmin == k] 
        ref[sel] = ave(ref[sel])
      }
    }
    
    # Re-order ref and replace in objectNorm
    objectNorm[,i] = ref[rmin]
  }
  
  results <- new("qsmooth")
  results@qsmoothWeights <- w
  
  rownames(objectNorm) = rownames(object)
  colnames(objectNorm) = colnames(object)
  results@qsmoothData <- objectNorm
  
  return(results)
}

R/AllClasses.R
#' @title the qsmooth class
#'
#' @description  Objects of this class store all 
#' the values needed information to work with a
#' qsmooth object
#' 
#' @slot qsmoothWeights qsmooth weights
#' @slot qsmoothData qsmooth normalized data
#' 
#' @return \code{qsmoothWeights} returns the qsmooth weights and 
#' \code{qsmoothData} returns the qsmooth normalized data 
#' 
#' @name qsmooth-class
#' @import methods
#' @exportClass qsmooth
#' @aliases qsmooth-class
#'
#' @examples
#' dat <- cbind(matrix(rnorm(1000), nrow=100, ncol=10), 
#'              matrix(rnorm(1000, .1, .7), nrow=100, ncol=10))
#' dat_qs <- qsmooth(object = dat, 
#'                   group_factor = rep(c(0,1), each=10))
#'
setClass(
    Class = "qsmooth", 
      slot = list(
        qsmoothWeights = "numeric", qsmoothData = "matrix")
)

#' @importFrom utils head
#' @importFrom utils tail
setMethod("show", "qsmooth", 
          function(object){
            cat("qsmooth: Smooth quantile normalization\n")
            cat("   qsmoothWeights (length =", 
                length(object@qsmoothWeights), "):", "\n")
            cat(c(head(round(object@qsmoothWeights,3), n=3), "...", 
                   tail(round(object@qsmoothWeights,3), n=3)), "\n")
            cat("   qsmoothData (nrows =", 
                dim(object@qsmoothData)[1], ", ncols =", 
                dim(object@qsmoothData)[2], "):", "\n")
            print(head(object@qsmoothData))
            if(dim(object@qsmoothData)[1] > 6){
              cat(" ....... \n") }
          }
)


R/qsmoothGC.R
.smooth_gcqn <- function(counts, 
                         gc,
                         group_factor, 
                         nGroups=50,
                         round=TRUE,
                         ...){
  gc_groups <- Hmisc::cut2(gc, g = nGroups)
  gcBinNormCounts <- matrix(NA, nrow=nrow(counts), ncol=ncol(counts), 
                            dimnames=list(rownames(counts),colnames(counts)))
  for(ii in seq_len(nlevels(gc_groups))){
    id <- which(gc_groups==levels(gc_groups)[ii])
    countBin <- counts[id,]
    qs <- qsmooth::qsmooth(countBin, group_factor=group_factor, ...)
    normCountBin <- qsmooth::qsmoothData(qs)
    if(round) normCountBin <- base::round(normCountBin)
    normCountBin[normCountBin<0] <- 0
    gcBinNormCounts[id,] <- normCountBin
  }
  return(gcBinNormCounts)
}


#' @title qsmoothGC
#' 
#' @description This function applies smoothed quantile 
#' normalization separately for groups of features that are
#' binned according to their GC-content.
#'
#' @param object an object which is a \code{matrix} or 
#' \code{data.frame} with observations (e.g. probes or genes) on 
#' the rows and samples as the columns. Alternatively, 
#' a user can provide a \code{SummarizedExperiment} object
#' and the \code{assay(object, "counts")} will be used as input 
#' for the qsmooth normalization.
#' @param gc GC-content of the features, ordered according to the features 
#' in \code{object}.
#' @param group_factor a group level continuous or categorial 
#' covariate associated with each sample or column in the 
#' \code{object}. The order of the \code{group_factor} must 
#' match the order of the columns in \code{object}.
#' @param nGroups The number of equally-sized bins used to group the 
#' GC-content values. Groups are created using \code{Hmisc::cut2}.
#' @param round Should normalized values be rounded to integers?
#' @param ... (Optional) Additional arguments passed to \code{\link{qsmooth}}.
#' @export 
#' 
#' @return A matrix of normalized counts.
#' 
#' @aliases qsmoothGC
#' 
#' @docType methods
#' @name qsmoothGC
#' @importFrom sva ComBat
#' @importFrom SummarizedExperiment assays
#' @importFrom SummarizedExperiment assay
#' @importFrom stats ave
#' @importFrom Hmisc cut2
#' 
#' @examples
#' dat <- cbind(matrix(rnorm(1000), nrow=100, ncol=10), 
#'              matrix(rnorm(1000, .1, .7), nrow=100, ncol=10))
#' gc <- runif(n=100, min=0.2, max=0.9)
#' dat_qs <- qsmoothGC(object = dat, 
#'                    gc = gc,
#'                    group_factor = rep(c(0,1), each=10))
#' 
#' @references 
#' Van den Berge K., Chou H., Roux de Bézieux H., Street K., Risso D., Ngai J., Dudoit S. Normalization benchmark of ATAC-seq datasets shows the 
#' importance of accounting for GC-content effects. https://www.biorxiv.org/content/10.1101/2021.01.26.428252v2
#' 
#' @rdname qsmoothGC
#' @export
qsmoothGC <- function(object, 
                      group_factor,
                      gc,
                      nGroups = 50,
                      round = TRUE,
                      ...){
  
  if(!any(is(object, "matrix") | is(object, "data.frame") |
          is(object, "SummarizedExperiment"))){
    stop("The class of the object must be a matrix, 
         data.frame or SummarizedExperiment")
  }
  
  if(is.data.frame(object)){ object <- as.matrix(object) }
  
  if(is(object, "SummarizedExperiment")){ 
    if("counts" %in% names(assays(object))){ 
      object <- assay(object, "counts")
    } else {
      stop("There is no assay slot named 'counts' inside 
           the object. Please check the names of the 
           assay slots using names(assays(object)).")
    }
  }
  
  objectNorm <- .smooth_gcqn(counts = object,
                             gc = gc,
                             group_factor = group_factor,
                             nGroups = nGroups, 
                             round = round,
                             ...)
  results <- new("qsmooth")
  results@qsmoothData <- objectNorm
  return(results)
}